### PR TITLE
Expand on best practice areas

### PR DESCRIPTION
**Pull request by iawells**
Fix up the titles so that they're more consistently using the same part
of speech.

Add a paragraph of description to each section.

Reorder the sections slightly ('install/start' comes before
'configure/manage' in a real life schenario; no reason the sections
should be the other way around).

(I note that these areas might also be areas for use cases.)

### CODE REVIEW

**Review by jeffsaelens — APPROVED** [review]
LGTM

File: doc/best_cnf_dev.md [hunk 1]
```diff
@@ -1,19 +1,25 @@
 # Best Practices for CNF Developers
 
 ## Table of Contents
-* [1.0 Configuration and Lifecycle](#1.0)
-* [2.0 Installable and Upgradeable](#2.0)
+* [1.0 Installation and Upgrade](#1.0)
+* [2.0 Configuration and Lifecycle](#2.0)
 * [3.0 Hardware support](#3.0)
 * [4.0 Microservice](#4.0)
```
Comment (iawells) [inline doc/best_cnf_dev.md]:
```suggestion
* [4.0 Microservices](#4.0)
```

File: doc/best_cnf_dev.md [hunk 2]
```diff
@@ -1,19 +1,25 @@
...
 * [5.0 Compatiblity](#5.0)
 * [6.0 State](#6.0)
 * [7.0 Security](#7.0)
 * [8.0 Scaling](#8.0)
-* [9.0 Observability](#9.0)
+* [9.0 Observability and Diagnostics](#9.0)
 
 <a name="1.0"></a>
-## 1.0 Configuration and Lifecycle
+## 2.0 Installation and Upgrade
```
Comment (taylor) [inline doc/best_cnf_dev.md]:
```suggestion
## 1.0 Installation and Upgrade
```

**Review by taylor — CHANGES_REQUESTED** [review]

**Review by CsatariGergely — APPROVED** [review]

**Review by xmulligan — APPROVED** [review]

**Review by taylor — APPROVED** [review]
lgtm

**Review by CsatariGergely — APPROVED** [review]

File: doc/best_cnf_dev.md [hunk 3]
```diff
@@ -33,49 +37,37 @@
...
-
-
+How to ensure CNFs work on as many platforms, in as many places, as possible; how to ensure that cloud platforms enable as many CNFs as possible; how to make CNFs work with each other and common elements in the operational stack.
 
 <a name="6.0"></a>
 ## 6.0 State
 
-
-
+Management of short-lived and long lived state, including state associated with network flows, configuration data, billing records and other data according to the varying requirements of resiliency, rate of change, access rate and persistence.
```
Comment (sishbi) [inline doc/best_cnf_dev.md]: Can 'billing records' be generalised to something like 'subscriber activity data'

**Review by sishbi — CHANGES_REQUESTED** [review]

File: doc/best_cnf_dev.md (hunk 3 above)
Comment (iawells) [reply doc/best_cnf_dev.md]: Yes, certainly.  Can you propose a change I can accept? (The doc icon with '+-' on it in the github comment toolbar)

Comment (sishbi) [reply doc/best_cnf_dev.md]: OK, I will try to find the button and then submit a PR on your PR (or whatever the correct term is)

Comment (iawells) [reply doc/best_cnf_dev.md]: It should just be a change as a part of a comment on this pull request.

Screenshot 2021-04-09 at 09 13 03

(See the icon above the word 'comment' in the screenshot?  If not, we may need to add a permission bit for you.)

File: doc/best_cnf_dev.md [hunk 4]
```diff
@@ -33,49 +37,37 @@
...
 
-
-
+Running of CNFs at a variety of different scales to manage different traffic requirements and subscriber counts.  Changing the scale of CNFs.
 
 <a name="9.0"></a>
-## 9.0 Observability
+## 9.0 Observability and Diagnostics
+
+How to get critical data about when things are going wrong and how to determine what must be done to put them right.  Using logs and metrics from all components in the system to narrow down the area where a problem exists, and to drill down into that area to determine a root cause and a fix.
```
Comment (Smitholi67) [inline doc/best_cnf_dev.md]: Is it fair to aspire for most of this to be automated?  If so, maybe it can read:

How to get critical data about when things are going wrong and to determine what must be done to put them right with limited or no human intervention.  Using logs and metrics from all components in the system to narrow down the area where a problem exists, determine root cause, and correct.

Comment (iawells) [reply doc/best_cnf_dev.md]: I see your point and I agree.  I might reword it to make it more that this is consumable by an automated system or a human being than prioritise automation over humans.  We have to serve both audiences anyway.

I don't disagree with you that automation is actually more important, but the humans involved sometimes do :)

Comment (iawells) [reply doc/best_cnf_dev.md]:
```suggestion
How to get critical data about when things are going wrong and how to determine what must be done to put them right.  The detection and correction may be through the actions of an operator or via an automated system.  Using logs and metrics from all components in the system to narrow down the area where a problem exists, and to drill down into that area to determine a root cause and a fix.
```

File: doc/best_cnf_dev.md (hunk 3 above)
Comment (iawells) [reply doc/best_cnf_dev.md]: (If this is a problem don't spend much time on it; just tell me it's all screwed up for you and I'll propose the fix.)

File: doc/best_cnf_dev.md (hunk 4 above)
Comment (Smitholi67) [reply doc/best_cnf_dev.md]: Thanks, this looks good to me.

**Review by Smitholi67 — APPROVED** [review]

File: doc/best_cnf_dev.md (hunk 3 above)
Comment (sishbi) [reply doc/best_cnf_dev.md]:
```suggestion
Management of short-lived and long lived state, including state associated with network flows, configuration data, subscriber activity data and other data according to the varying requirements of resiliency, rate of change, access rate and persistence.
```

**Review by sishbi — APPROVED** [review]

File: doc/best_cnf_dev.md (hunk 3 above)
Comment (sishbi) [reply doc/best_cnf_dev.md]: @iawells I think I did the right thing, let me know if not...

Comment (iawells) [reply doc/best_cnf_dev.md]: Perfect, thanks.  I'll talk to @taylor and @jeffsaelens about adding this to the reviewer help.
